packages/fetchai/skills/tac_participation_contract/handlers.py

```diff
@@ -0,0 +1,517 @@
+                if tac_msg_type == TACMessage.Type.TRANSACTION_CONFIRMATION:
+                    self._on_transaction_confirmed(tac_msg)
+                elif tac_msg_type == TACMessage.Type.CANCELLED:
+                    self._on_cancelled()
+                # elif tac_msg_type == TACMessage.Type.STATE_UPDATE:
+                #     self._on_state_update(tac_msg, sender)
+            elif game.phase.value == Phase.POST_GAME.value:
+                raise ValueError(
+                    "We do not expect a controller agent message in the post game phase."
+                )
+        except ValueError as e:
+            self.context.logger.warning(str(e))
+
+    def teardown(self) -> None:
+        """
+        Implement the handler teardown.
+
+        :return: None
+        """
+        pass
+
+    def _on_tac_error(self, tac_message: TACMessage) -> None:
+        """
+        Handle 'on tac error' event emitted by the controller.
+
+        :param tac_message: The tac message.
+
+        :return: None
+        """
+        error_code = tac_message.error_code
+        self.context.logger.error(
+            "[{}]: Received error from the controller. error_msg={}".format(
+                self.context.agent_name, TACMessage._from_ec_to_msg.get(error_code)
+            )
+        )
+        if error_code == TACMessage.ErrorCode.TRANSACTION_NOT_VALID:
+            info = tac_message.info
+            transaction_id = (
+                cast(str, info.get("transaction_id"))
+                if (info.get("transaction_id") is not None)
+                else "NO_TX_ID"
+            )
+            self.context.logger.warning(
+                "[{}]: Received error on transaction id: {}".format(
+                    self.context.agent_name, transaction_id[-10:]
+                )
+            )
+
+    def _on_start(self, tac_message: TACMessage) -> None:
+        """
+        Handle the 'start' event emitted by the controller.
+
+        :param tac_message: the game data
+
+        :return: None
+        """
+        self.context.logger.info(
+            "[{}]: Received start event from the controller. Starting to compete...".format(
+                self.context.agent_name
+            )
+        )
+        game = cast(Game, self.context.game)
+        game.init(tac_message, tac_message.counterparty)
+        game.update_game_phase(Phase.GAME)
+        amount_by_currency_id = {
+            str(key): value for key, value in tac_message.amount_by_currency_id.items()
+        }
+        quantities_by_good_id = {
+            str(key): value for key, value in tac_message.quantities_by_good_id.items()
+        }
+        exchange_params_by_currency_id = {
+            str(key): value
+            for key, value in tac_message.exchange_params_by_currency_id.items()
+        }
+        utility_params_by_good_id = {
+            str(key): value
+            for key, value in tac_message.utility_params_by_good_id.items()
+        }
```

Comment:
The main change is here that I am changing the type of the keys in the dictionaries. For some reason I receive them as ints. Thought I am investigating in order to remove this conversion.